Pushrules for relations (#3664)
matrix-org/matrix-doc: proposals/3664-notifications-for-relations.md

largely similar to the existing `event_match`, but operates on the related-to
event. Such a condition could look like this:

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.in_reply_to",
  "key": "sender",
  "pattern": "@me:my.server"
}
```

This condition can be used to notify me whenever someone sends a reply to my
messages.

- `rel_type` is the relation type. For the sake of compatibility replies
    should be matched as if they were sent in the relation format from
    [MSC2674](https://github.com/matrix-org/matrix-doc/pull/2674) with a
    `rel_type` of `m.in_reply_to`. If the event has any relation of this type,
    the related event should be matched using `pattern` and `key`.
- `key` (optional): The dot-separated field of the event to match, e.g. `content.body`
    or `sender`. If it is not present, the condition should match all events,
    that have a relation of type `rel_type`.
- `pattern` (optional): The glob-style pattern to match against. Patterns with
    no special glob characters should be treated as having asterisks prepended
    and appended when testing the condition. If this is not present, it should

>     match everything if the specific key is present.

Reviewer: so... are you proposing that the behaviour should match what the spec currently (incorrectly) says for `event_match`, or the actual behaviour for `event_match`?

>     match everything if the specific key is present.

Author: What the actual behaviour is, I had tried to clarify that with https://github.com/matrix-org/matrix-doc/pull/3664/files#diff-0ab594b71fe15d3959c5ee0d56c79b7c111d98ee81859831140a4e92d5114c04R53

>     match everything if the specific key is present.

Reviewer: oh I see. I think it would be clearer to just update this text to match the actual behaviour.

>     match everything if the specific key is present.

Author: I just removed the description of the behaviour, said it is the same as for event_match and linked to the issue for the curious. That's easier than trying to describe the current behaviour exactly :D

>     match everything if the specific key is present.

Reviewer: it certainly is.

> `key` and `pattern` are optional to allow you to enable or suppress all

Author: I guess it makes the most sense to put it with `.m.rule.contains_display_name` then

> `key` and `pattern` are optional to allow you to enable or suppress all

Author: I think the only issue is, when there is a pattern specified, but no `key`. It makes sense to reject that, when setting such a rule, but since clients can't do that, they should just ignore the pattern in that case, imo.

notifications for a specific event type. For example one could suppress
notifications for all events in

> [threads](https://github.com/matrix-org/matrix-doc/pull/3440) and all

Reviewer 2: I think it's worth describing how threads specifically works in this MSC. Through discussions OOB, it's important for threads to *not* cause mentions just because they use the fallback, *unless* the relationship is also `m.render_in: [m.thread]` in which case it's a reply in a thread that *should* cause a mention (just like a regular reply).

Author: I've added a few examples of how the threading MSC could define push rules to get the notification behaviour they want. I think it is best to define the rule in that MSC though and there are a few difficulties, that could be better solved by modifying the threading MSC.

Reviewer 2: I don't think it's the best idea to overload the threading MSC with its push rules and would rather include them here, as this stream of working is being driven by threads at the moment.

If it's not comfortable for inclusion here then we can open a new MSC which considers both cases.

Author: I think notification behaviour for threads should be its own MSC. Threads are using replies as a fallback, which I don't think is a great idea, add metadata to the replies to then disable rendering of those fallbacks conditionally and I don't think that is even finalized yet. I also don't have any experience with threads and notifications for threads are very complicated to get right, because you pretty much don't want to mark all messages in a thread as read, when opening one of them. I think all of that should be figured out in one MSC and I don't want to block this MSC on that, because threads  have a lot more issues than replies. The threading MSC also changed its fallback twice already, where it completely dropped it in the middle.

So yeah, I think pushrules for threads need someone working on threads to look at, not someone like me, who actively avoids them. In the best case you just extend the event_match rule to allow inverted matches, so that you can enable the reply rule only if there is no thread relation and then add an extra rule for replies in threads.

Reviewer 3: Okay, to unbreak the stalemate on this one, i suggest that the thread-specific notification behaviour (i.e. how to configure push rules for threads, and whether we need to change push_rule syntax to support negations etc) is split into a separate MSC.  Currently this MSC reads a bit weirdly in that it refers to threads a bunch and then proposes a partial solution - either it needs to flesh it out entirely or split it off into a separate MSC.

Author: Yeah, I think that makes sense. I mostly just wanted to show how to mute specific threads as examples that were possible with this rule type. I didn't want to specify any specific rules for threads, because I just lack enough knowledge about them to know, what a good solution would be. I might be able to specify something, if I had a call with someone working on threads, but I think it is best if someone working on threads defines the notification behaviour, that they are looking for.

Author: I removed the proposal for threads now. I just kept the one thread rule as an example, because it just shows, how the rules can be used without defining any actual behaviour. I think that should be fine and I hope it isn't misunderstood as an actual rule but rather as an example of useful applications of this condition kind.

Reviewer: I think this is clear now, thanks to #3676

two conditions:

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.replace"
}
```

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.thread"
}
```

Without a `key` the push rule can be evaluated without fetching the related to
event.

### A push rule for replies

To enable notifications for replies without relying on the reply fallback, but
with similar semantics we also define a new default push rule. The proposed
push rule differs slightly from the old behaviour, because it only notifies you
for replies to your events, but it does not notify you for replies to events,
which contained your display name or matrix ID. The rule should look like this:

```json5
{
    "rule_id": ".m.rule.reply",
    "default": true,
    "enabled": true,
    "conditions": [
        {
            "kind": "related_event_match"
            "rel_type": "m.in_reply_to",
            "key": "sender",
            "pattern": "[the user's Matrix ID]"
        }
    ],
    "actions": [
        "notify",
        {
            "set_tweak": "sound",
            "value": "default"
        },
        {
            "set_tweak": "highlight"
        }
    ]
}
```

> This should be an underride rule, since it can't be a content rule. It ensures

Author: I guess it makes the most sense to put it with `.m.rule.contains_display_name` then

you get notified for replies to all events you sent. The actions are the same as
for `.m.rule.contains_display_name` and `.m.rule.contains_user_name`.

No other rules are proposed as no other relations are in the specification as of
writing this MSC to decrease dependencies.

> ## Potential issues

Reviewer: Note that clients which support configuring notifications will need to be prepared to deal with servers which do not yet support the new rule. I think they can infer this from the presence of  `.m.rule.reply` in the response to `/pushrules`?

Reviewer 4: Worse case clients could just handle the PUT failing, but that's not ideal to show an option that users don't know they can't use until they make the change.

Author: Couldn't the `.m.rule.reply` be deleted by a user/client, which would then make a client think, that relation rules are unsupported? I think checking for existence of that rule only works during the transition period, after that clients should just check for the client API version maybe? I.e. on 1.2 or 1.3 clients could just assume it is supported even if the reply rule is not present? It leads to a small UX issue for servers claiming support for that API version, but not supporting the condition, but I hope that is a very small percentage of users.

Reviewer: > Couldn't the .m.rule.reply be deleted by a user/client, which would then make a client think, that relation rules are unsupported? 

default rules cannot be deleted, only disabled.

Author: Added this now.

Reviewer: > Added this now.

sorry, I'm not able to see the backwards-compatibility problem for clients being spelt out as an issue?

Author: https://github.com/matrix-org/matrix-doc/pull/3664/files#diff-0ab594b71fe15d3959c5ee0d56c79b7c111d98ee81859831140a4e92d5114c04R157 is this not it?

Reviewer: I don't think so.

Let me reiterate:

> **clients which support** configuring [the new push rule] will need to be prepared to deal with **servers which do not yet support** the new rule.

Author: I've added a capabilities flag now. If clients want to enable that rule, when the server doesn't support them, can be up to them :3

- Most push rules for relations will need a lookup into a second event. This
    causes additional implementation complexity and can potentially be
    expensive. Looking up one event shouldn't be that heavy, but it is overhead,
    that wasn't there before and it needs to be evaluated for every event, so it
    clearly is somewhat performance sensitive.
- If the related to event is not present on the homeserver, evaluating the
    push rule may be delayed or fail completely. For most rules this should not
    be an issue. You can assume the event was not sent by a user on your server

>     if the event is not present on your server.

Reviewer 4: Yup, ran into this in our implementation in matrix-react-sdk, the code there is currently not async so it best effort attempts to use whatever is in the in-memory stores and then just fails to evaluate if the event isn't present locally. Will throw up a branch on github for that work and link it to this MSC as well.

# MSC3664: Notifications for relations

Relations are very powerful and are becoming a platform to build new features
like replies, edits, reactions, threads, polls and much more on.

> On the other hand there is currently no way to control, what you are getting

Reviewer: ```suggestion
On the other hand there is currently no way to control what you are getting
```

> notified for. Some people want to get notified, when someone replies to their

Reviewer: ```suggestion
notified for. Some people want to get notified when someone replies to their
```

for replies to your events, but it does not notify you for replies to events,
which contained your display name or matrix ID. The rule should look like this:

```json5
{
    "rule_id": ".m.rule.reply",
    "default": true,
    "enabled": true,
    "conditions": [
        {
            "kind": "related_event_match",
            "rel_type": "m.in_reply_to",
            "key": "sender",
            "pattern": "[the user's Matrix ID]"
        }
    ],
    "actions": [
        "notify",
        {
            "set_tweak": "sound",
            "value": "default"
        },
        {
            "set_tweak": "highlight"
        }
    ]
}
```

This should be an override rule, since it can't be a content rule and should
not be overridden when setting a room to mentions only. It should be placed just
before `.m.rule.contains_display_name` in the list. This ensures you get
notified for replies to all events you sent. The actions are the same as for
`.m.rule.contains_display_name` and `.m.rule.contains_user_name`.

No other rules are proposed as no other relations are in the specification as of
writing this MSC to decrease dependencies.

## Potential issues

> Most push rules for relations will need a lookup into a second event. This

Reviewer: It might be helpful to number each separate issue.

causes additional implementation complexity and can potentially be expensive.

> Looking up one event shouldn't be that heavy, but it is overhead, that wasn't

Reviewer: ```suggestion
Looking up one event shouldn't be that heavy, but it is overhead that wasn't
```

rule may be delayed or fail completely. For most rules this should not be an
issue. You can assume the event was not sent by a user on your server if the
event is not present on your server.  In general clients and servers should do
their best to evaluate the condition. If they fail to do so (possibly because
they can't look up the event asynchronously) in a timely manner, the condition
may be ignored/evaluated to false. This should affect only a subset of events,
because in general relations happen to events in close proximity. There is a
risk of missing notifications for replies to very old messages and similar
relations.

[Threads](https://github.com/matrix-org/matrix-doc/pull/3440) use replies
[as a fallback](https://github.com/matrix-org/matrix-doc/pull/3440/files#diff-113727ce0257b4dc0ad6f1087b6402f2cfcb6ff93272757b947bf1ce444056aeR82).
This would cause a notification with the new `.m.rule.reply` rule. To prevent
that another MSC should add a or modify the existing push rules to not apply to
threads, when they are not a reply. This MSC does not spell out a solution for
that, because the author lacks familiarity with the design goals of the

> threading proposals.

Reviewer: this is no longer true - remove?

Author: How was that resolved? Did I miss something?

Reviewer: Threads are no longer using reply fallbacks, I believe, per msc3676

Author: No, threads still use the reply relation as a fallback. The above sentence is about the new reply rule, which triggers on reply relations, would cause notifications gor threads, because they send reply relations all the time. MSC3676 does not change that.

Reviewer: oh, crap, you're right.

Author: I resolved this now by adding an option to ignore fallback relations, although I noticed that the `is_falling_back` key is not actually on the relation, which is falling back. Apart from this the MSC should work fine now with threads.

> Adding a new rule, that causes notifications, forces users to change their

Reviewer: ```suggestion
Adding a new rule that causes notifications, forces users to change their
```

notification settings again. In this case, if the user disabled pings for
mentions, they will need to disable the rule manually again, if they did so to
disable replies. In the transition period clients might not have a UI to do so.
This is a risk with all push rule changes and since it allows for a much better

> control over what notifies you, the trandeoff should be acceptable. Many users

Reviewer: ```suggestion
control over what notifies you, the tradeoff should be acceptable. Many users
```

notified for edits, others prefer to not get notified, when someone fixes typos
20 times in a row for a long message they sent a week ago.

We should extend push rules so that a server can provide sane defaults and users
can adjust them to their own wishes.

## Proposal

### New push rule condition: `related_event_match`

Notifications for relation based features need to distinguish, what type of
relation was used and potentially match on the content of the related-to event.

To do that we introduce a new type of condition: `related_event_match`. This is
largely similar to the existing `event_match`, but operates on the related-to
event. Such a condition could look like this:

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.in_reply_to",
  "key": "sender",
  "pattern": "@me:my.server"
}
```

This condition can be used to notify me whenever someone sends a reply to my
messages.

- `rel_type` is the relation type. For the sake of compatibility replies
    should be matched as if they were sent in the relation format from
    [MSC2674](https://github.com/matrix-org/matrix-doc/pull/2674) with a
    `rel_type` of `m.in_reply_to`. If the event has any relation of this type,

>     the related event should be matched using `pattern` and `key`.

Reviewer 5: This seems to implicitly usurp a relation type of `m.in_reply_to`, which might not be desired? I think there's been some thought these should be replaced eventually with MSC3267 (`m.reference` relations).

Regardless, it means that `m.in_reply_to` could not be used for a relation in the future and would need to be reserved.

Reviewer 5: I still find it a bit hidden in this MSC that this grabs the `m.in_reply_to` as if it is a releation type when it isn't. I think this will be confusing for implementers (and difficult to specify properly).

>     the related event should be matched using `pattern` and `key`.

Author: It should be matched as the existing events do using the match condition, just in this case it operates on the related event, yes. I'll try to clarify that a bit.

>     the related event should be matched using `pattern` and `key`.

Author: Should be clearer now.

## Proposal

### New push rule condition: `related_event_match`

Notifications for relation based features need to distinguish what type of
relation was used and potentially match on the content of the related-to event.

To do that we introduce a new type of condition: `related_event_match`. This is
largely similar to the existing `event_match`, but operates on the related-to
event. Such a condition could look like this:

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.in_reply_to",
  "include_fallbacks": false,
  "key": "sender",
  "pattern": "@me:my.server"
}
```

This condition can be used to notify me whenever someone sends a reply to my
messages.

- `rel_type` is the relation type. For the sake of compatibility, replies
    should be matched as if they were sent in the relation format from
    [MSC2674](https://github.com/matrix-org/matrix-doc/pull/2674) with a
    `rel_type` of `m.in_reply_to`. If the event has any relation of this type,
    the related event should be matched using `pattern` and `key`.
- `include_fallbacks` decides if the relation should be followed even for
    fallbacks (i.e. relations with the `is_falling_back` property set to `true`

>     like for threads). Defaults to false so only actual relations are counted.

Reviewer: ```suggestion
    like for [threads](https://github.com/matrix-org/matrix-spec-proposals/blob/main/proposals/3440-threading-via-relations.md#backwards-compatibility).
    Defaults to `false` so only actual relations are counted.
```

- `key` (optional): The dot-separated field of the event to match, e.g. `content.body`
    or `sender`. If it is not present, the condition should match all events,
    that have a relation of type `rel_type`.
- `pattern` (optional): The glob-style pattern to match against.

> `key` and `pattern` have exactly the same meaning as in `event_match`
> conditions. See https://github.com/matrix-org/matrix-doc/issues/2637 for a
> clarification of their behaviour.

Reviewer: Now that https://github.com/matrix-org/matrix-spec-proposals/issues/2637 is fixed, I think we can omit this link and just do:

```suggestion
`key` and `pattern` have exactly the same meaning as in 
[`event_match` conditions](https://spec.matrix.org/unstable/client-server-api/#conditions-1).
```

`key` and `pattern` are optional to allow you to enable or suppress all
notifications for a specific relation type. For example one could suppress
notifications for all events with a relation from
[threads](https://github.com/matrix-org/matrix-doc/pull/3440) and all
[edits](https://github.com/matrix-org/matrix-doc/pull/2676) with the following
two conditions:

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.replace"
}
```

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.thread"
}
```

Without a `key` and `pattern` the push rule can be evaluated without fetching
the related to event. If one of those two fields is missing, a server should

> prevent those rules from being added with the appropriate error code. (A client
> wouldn't have a choice but to ignore those keys if the server failed to prevent
> the rule from being added.)

Reviewer: > A client wouldn't have a choice

I don't think this is true. Another, equally valid choice, is to ignore the entire push rule.

```suggestion
prevent those rules from being added with the appropriate error code. (A client
which sees a `related_event_match` condition with one, but not both, of `key` and `pattern` should
ignore the `key`/`pattern` property.
```

writing this MSC to decrease dependencies.

## Potential issues

Most push rules for relations will need a lookup into a second event. This
causes additional implementation complexity and can potentially be expensive.
Looking up one event shouldn't be that heavy, but it is overhead that wasn't
there before and it needs to be evaluated for every event, so it clearly is
somewhat performance sensitive.

If the related to event is not present on the homeserver, evaluating the push
rule may be delayed or fail completely. For most rules this should not be an
issue. You can assume the event was not sent by a user on your server if the
event is not present on your server.  In general clients and servers should do
their best to evaluate the condition. If they fail to do so (possibly because
they can't look up the event asynchronously) in a timely manner, the condition
may be ignored/evaluated to false. This should affect only a subset of events,
because in general relations happen to events in close proximity. There is a
risk of missing notifications for replies to very old messages and similar
relations.

[Threads](https://github.com/matrix-org/matrix-doc/pull/3440) use replies

> [as a fallback](https://github.com/matrix-org/matrix-doc/pull/3440/files#diff-113727ce0257b4dc0ad6f1087b6402f2cfcb6ff93272757b947bf1ce444056aeR82).

Reviewer: This link doesn't seem terribly helpful. I think this would be more useful:

```suggestion
[as a fallback](https://github.com/matrix-org/matrix-spec-proposals/blob/main/proposals/3440-threading-via-relations.md#backwards-compatibility).
```

This would cause a notification with the new `.m.rule.reply` rule. To prevent
that this MSC adds the `include_fallbacks` key to the rule, so that reply
relations only added as a fallback are ignored. (Currently `is_falling_back` key
is in a bit of a weird location. Maybe this can be amended in the threading MSC

> to be a bit more generic before it is added to the spec.)

Reviewer: the threading MSC has landed, so changing it would mean a new MSC and changing a bunch of implementations. I suggest opening a separate issue for this explaining your concerns.

Author: Well, it is not really an issue, servers can work around that to carry the threading semantics forward or not. It does make the code quite ugly, but eh, that ship has sailed.

Reviewer 5: The `include_fallbacks` is on the event in question, not the related event, correct? I find it a bit weird that `related_event_match` matches some fields on the current event and other fields on the related event.

(Can we also update the links to point the spec?)

before `.m.rule.contains_display_name` in the list. This ensures you get
notified for replies to all events you sent. The actions are the same as for
`.m.rule.contains_display_name` and `.m.rule.contains_user_name`.

No other rules are proposed as no other relations are in the specification as of
writing this MSC to decrease dependencies.

## Potential issues

Most push rules for relations will need a lookup into a second event. This
causes additional implementation complexity and can potentially be expensive.
Looking up one event shouldn't be that heavy, but it is overhead that wasn't
there before and it needs to be evaluated for every event, so it clearly is
somewhat performance sensitive.

If the related to event is not present on the homeserver, evaluating the push
rule may be delayed or fail completely. For most rules this should not be an
issue. You can assume the event was not sent by a user on your server if the
event is not present on your server.  In general clients and servers should do
their best to evaluate the condition. If they fail to do so (possibly because
they can't look up the event asynchronously) in a timely manner, the condition

> may be ignored/evaluated to false. This should affect only a subset of events,

Reviewer: which? if we ignore the condition, it is effectively true. So should it be considered to be true or false?

Author: Right, the rule is ignored or the condition is evaluated to false, which should mean the same thing.

because in general relations happen to events in close proximity. There is a
risk of missing notifications for replies to very old messages and similar

> relations.

Reviewer: stuff like this makes me feel even more strongly that I want to see a client implementation before this can proceed. Is this tractable for a client? does skipping the condition if we don't have the target event give an acceptable UX?

    like for [threads](https://github.com/matrix-org/matrix-spec-proposals/blob/main/proposals/3440-threading-via-relations.md#backwards-compatibility).
    Defaults to `false` so only actual relations are counted.
- `key` (optional): The dot-separated field of the event to match, e.g. `content.body`
    or `sender`. If it is not present, the condition should match all events,
    that have a relation of type `rel_type`.
- `pattern` (optional): The glob-style pattern to match against.

`key` and `pattern` have exactly the same meaning as in 
[`event_match` conditions](https://spec.matrix.org/unstable/client-server-api/#conditions-1).

`key` and `pattern` are optional to allow you to enable or suppress all
notifications for a specific relation type. For example one could suppress
notifications for all events with a relation from
[threads](https://github.com/matrix-org/matrix-doc/pull/3440) and all
[edits](https://github.com/matrix-org/matrix-doc/pull/2676) with the following
two conditions:

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.replace"
}
```

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.thread"
}

> ```

Reviewer 5: Aren't these identical to:

```json
{
  "kind": "event_match",
  "key": "content.m\\.relates_to.\\.rel_type",
  "pattern": "m.replace"
}
```

(And a corresponding one for `m.thread`.)

I'm not sure this example really shows the need for this MSC?

Author: Well, in the case of replies you need to also match, if it isn't a fallback. Also, it interacts better with #3051 or any other reorganization of the relation format. Matching dotted keys is also #3873. Yes, you can solve this differently, but that solution also has downsides. This is just a convenient extension of this MSC, the actual meat is matching the content of the related event.

No other rules are proposed as no other relations are in the specification as of
writing this MSC to decrease dependencies.

## Potential issues

Most push rules for relations will need a lookup into a second event. This
causes additional implementation complexity and can potentially be expensive.
Looking up one event shouldn't be that heavy, but it is overhead that wasn't
there before and it needs to be evaluated for every event, so it clearly is
somewhat performance sensitive.

If the related to event is not present on the homeserver, evaluating the push
rule may be delayed or fail completely. For most rules this should not be an
issue. You can assume the event was not sent by a user on your server if the
event is not present on your server.  In general clients and servers should do
their best to evaluate the condition. If they fail to do so (possibly because
they can't look up the event asynchronously) in a timely manner, the rule
may be ignored/the condition evaluated to false. This should affect only a
subset of events, because in general relations happen to events in close
proximity. There is a risk of missing notifications for replies to very old

> messages and similar relations.

Reviewer 5: This doesn't really define the homeserver behavior here -- are they to attempt to backfill if they can't find the event? Just skip it?

Similarly for clients, should they attempt to ask the homeservers for the missing event if they don't have it locally?

Author: That's a quality of implementation issue and I don't think that needs to be defined. For example in the case of matching replies you usually already have the events sent from your server and don't need to backfill them. In threads this is similar, usually you have the participating thread already. If you don't have an event, it is a tradeoff. Do you want accurate notifications or low latency? Alternatively you can always update the notification counter asynchronously. I don't see a need to require any specific behaviour here.

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.in_reply_to",
  "key": "sender",
  "pattern": "@me:my.server"
}
```

This condition can be used to notify me whenever someone sends a reply to my
messages.

- `rel_type` is the relation type. For the sake of compatibility replies
    should be matched as if they were sent in the relation format from
    [MSC2674](https://github.com/matrix-org/matrix-doc/pull/2674) with a
    `rel_type` of `m.in_reply_to`. If the event has any relation of this type,
    the related event should be matched using `pattern` and `key`.
- `key` (optional): The dot-separated field of the event to match, e.g. `content.body`
    or `sender`. If it is not present, the condition should match all events,
    that have a relation of type `rel_type`.
- `pattern` (optional): The glob-style pattern to match against.

`key` and `pattern` have exactly the same meaning as in `event_match`
conditions. See https://github.com/matrix-org/matrix-doc/issues/2637 for a
clarification of their behaviour.

`key` and `pattern` are optional to allow you to enable or suppress all
notifications for a specific relation type. For example one could suppress

> notifications for all events with a realation from

Reviewer 3: ```suggestion
notifications for all events with a relation from
```

    no special glob characters should be treated as having asterisks prepended
    and appended when testing the condition. If this is not present, it should
    match everything if the specific key is present.

`key` and `pattern` have exactly the same meaning as in `event_match`
conditions, the wording is taken from the current spec. The wording of that is
currently debated in https://github.com/matrix-org/matrix-doc/issues/2637 and
this MSC just follows whatever the spec does for `event_match`.

`key` and `pattern` are optional to allow you to enable or suppress all
notifications for a specific event type. For example one could suppress
notifications for all events in
[threads](https://github.com/matrix-org/matrix-doc/pull/3440) and all
[edits](https://github.com/matrix-org/matrix-doc/pull/2676) with the following
two conditions:

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.replace"
}
```

```json5
{
  "kind": "related_event_match",
  "rel_type": "m.thread"
}
```

Without a `key` the push rule can be evaluated without fetching the related to
event. If only `key` is present, `pattern` should be assumed to be `*`, which

> allows you to match for a field being present regardless of its value. If only

Reviewer: this seems a bit random. Why the difference to `event_match`?

Author: Because I think it is useful to be able to match if a field exists without having to add a pattern. Especially since the pattern can be difficult, when it needs to match an array. But I can remove that, if it doesn't sound useful enough or should be done for both condition kinds at the same time.

Reviewer: I think it will be confusing to make it different to `event_match`, particularly given there is a way to achieve the behaviour you want with an explicit `*`.

> `value` is present, servers should return an error when setting the rule.

Reviewer: s/value/pattern?

> Clients should ignore the `pattern` field if there is no `key` field.

Reviewer: ... but then, how could there be a `pattern` with no `key`?

Author: If there was a bug in the server or whatever. I know the spec usually does not define behaviour for bugs, but a client simply can't prevent such a rule from being added by a different client, so I thought it would make sense to just define some guidance for how a client could handle such rules. Ideally that case just never happens.

### A push rule for replies

To enable notifications for replies without relying on the reply fallback, but
with similar semantics we also define a new default push rule. The proposed
push rule differs slightly from the old behaviour, because it only notifies you
for replies to your events, but it does not notify you for replies to events,
which contained your display name or matrix ID. The rule should look like this:

```json5
{
    "rule_id": ".m.rule.reply",
    "default": true,
    "enabled": true,
    "conditions": [
        {
            "kind": "related_event_match",
            "rel_type": "m.in_reply_to",
            "key": "sender",
            "pattern": "[the user's Matrix ID]"
        }
    ],
    "actions": [
        "notify",
        {
            "set_tweak": "sound",
            "value": "default"
        },
        {
            "set_tweak": "highlight"
        }
    ]
}
```

This should be an override rule, since it can't be a content rule and should

> not be overridden when setting a room to mentions only. It should be places just
> before `.m.rule.contains_display_name` in the list. tThis ensures you get

Reviewer 6: ```suggestion
not be overridden when setting a room to mentions only. It should be placed just
before `.m.rule.contains_display_name` in the list. This ensures you get
```

notified for replies to all events you sent. The actions are the same as for

> `.m.rule.contains_display_name` and `.m.rule.contains_user_name`.

Reviewer: so, suppose you've disabled notifications - or set them to be silent - for messages containing your username, which currently includes replies. Once this change is made, suddenly you will start receiving noisy notifications for replies. Worse, until clients are updated to support the new push rule, there is no way to turn them off!

Some people get quite upset when this sort of thing happens. I think it may be unavoidable, but let's at least call it out as an issue.

Author: I called it out and also tried to justify, why this should not be as bad. I think many people disable the mention based pings, because 2 users share the same name or similar mistaken pings. Those should be much less frequent with this rule, which should help mitigate this somewhat and it also won't ping you, if you don't write in that room at all, which is not the case for mentions. I hope that is okay.

No other rules are proposed as no other relations are in the specification as of
writing this MSC to decrease dependencies.

## Potential issues

Most push rules for relations will need a lookup into a second event. This
causes additional implementation complexity and can potentially be expensive.
Looking up one event shouldn't be that heavy, but it is overhead, that wasn't
there before and it needs to be evaluated for every event, so it clearly is
somewhat performance sensitive.

If the related to event is not present on the homeserver, evaluating the push
rule may be delayed or fail completely. For most rules this should not be an
issue. You can assume the event was not sent by a user on your server if the
event is not present on your server.  In general clients and servers should do
their best to evaluate the condition. If they fail to do so (possibly because
they can't look up the event asynchronously) in a timely manner, the condition
may be ignored/evaluated to false. This should affect only a subset of events,
because in general relations happen to events in close proximity. There is a
risk of missing notifications for replies to very old messages and similar
relations.

> [threads](https://github.com/matrix-org/matrix-doc/pull/3440) use replies

Reviewer: ```suggestion
[Threads](https://github.com/matrix-org/matrix-doc/pull/3440) use replies
```
